mqtt/mqtt3/src/client/subscriptions.rs

```diff
@@ -32,119 +32,117 @@ impl State {
-                    // Instead, if there's an error, we'll update self.subscriptions anyway with the expected QoS, and set the error to be returned here.
-                    // The error will reset the session and resend the subscription requests, including these that didn't match the expected QoS,
-                    // so pretending the subscription succeeded does no harm.
-                    let mut err = None;
-                    for (
-                        crate::proto::SubscribeTo {
-                            topic_filter,
-                            qos: expected_qos,
-                        },
-                        qos,
-                    ) in subscribe_to.into_iter().zip(qos)
-                    {
-                        match qos {
-                            crate::proto::SubAckQos::Success(actual_qos) => {
-                                if actual_qos >= expected_qos {
-                                    log::debug!(
-                                        "Subscribed to {} with {:?}",
-                                        topic_filter,
-                                        actual_qos
-                                    );
-                                    self.subscriptions.insert(topic_filter.clone(), actual_qos);
-                                    subscription_updates.push(
-                                        super::SubscriptionUpdateEvent::Subscribe(
-                                            crate::proto::SubscribeTo {
-                                                topic_filter,
-                                                qos: actual_qos,
-                                            },
-                                        ),
-                                    );
-                                } else {
-                                    if err.is_none() {
-                                        err = Some(super::Error::SubscriptionDowngraded(
-                                            topic_filter.clone(),
-                                            expected_qos,
-                                            actual_qos,
-                                        ));
-                                    }
+                        packet_identifiers.discard(packet_identifier);
 
-                                    self.subscriptions.insert(topic_filter, expected_qos);
+                        // We can't put subscribe_to back into self.subscription_updates_waiting_to_be_acked within the below loop
+                        // since we would've partially consumed it.
+                        // Instead, if there's an error, we'll update self.subscriptions anyway with the expected QoS, and set the error to be returned here.
+                        // The error will reset the session and resend the subscription requests, including these that didn't match the expected QoS,
+                        // so pretending the subscription succeeded does no harm.
```

Comment:
We need to update this TODO right?

Reply:
added a comment where it returns the event, I think this part still makes sense for other cases